framework/finalizer.go

```diff
@@ -0,0 +1,116 @@
+package framework
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	finalizerPrefix = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
+	}
+	finalizerName := getFinalizerName(f.name)
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // object already has the finalizer.
+	}
+
+	patch := patchSpec{
+		Op:    "add",
+		Value: finalizerName,
+		Path:  "/metadata/finalizers/-",
+	}
+	p, err := json.Marshal(patch)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	operation := func() error {
+		res := f.restClient.Patch(types.JSONPatchType).AbsPath(accessor.GetSelfLink()).Body(p).Do()
+		if res.Error() != nil {
+			return microerror.Mask(res.Error())
+		}
+		return nil
+	}
+	err = backoff.Retry(operation, f.backOffFactory())
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	return true, nil
+}
+
+func (f *Framework) removeFinalizer(ctx context.Context, obj interface{}) error {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	finalizerName := getFinalizerName(f.name)
+	if !containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		f.logger.LogCtx(ctx, "function", "removeFinalizer", "level", "warning", "message", "object is missing a finalizer")
+		return nil // object has no finalizer, probably migration.
+	}
+
+	patch := patchSpec{
+		Op:    "replace",
+		Value: removeFinalizer(accessor.GetFinalizers(), finalizerName),
+		Path:  "/metadata/finalizers",
+	}
+	p, err := json.Marshal(patch)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	operation := func() error {
+		res := f.restClient.Patch(types.JSONPatchType).AbsPath(accessor.GetSelfLink()).Body(p).Do()
+		if res.Error() != nil {
+			return microerror.Mask(res.Error())
+		}
+		return nil
+	}
+	err = backoff.Retry(operation, f.backOffFactory())
```

Comment:
I tried to add retries using the `backOffFactory` from the framework. Not sure if this is the way to go here.